From 3a87e4530d065a114024a602fa76a4b94d905801 Mon Sep 17 00:00:00 2001
From: Chase Douglas <chase.douglas@canonical.com>
Date: Tue, 29 Mar 2011 10:28:52 -0400
Subject: [PATCH 2/3] Use the number of active touches instead of BTN_*TAP

The kernel doesn't know about inactive areas of the touchpad, so it
doesn't filter out inactive touches when counting the number of fingers
for events like BTN_DOUBLETAP.

When touch data is provided, use the current active touch count instead.

Signed-off-by: Chase Douglas <chase.douglas@canonical.com>
---
 src/eventcomm.c |    8 +++++++-
 src/eventcomm.h |    1 +
 2 files changed, 8 insertions(+), 1 deletions(-)

Index: xserver-xorg-input-synaptics/src/eventcomm.c
===================================================================
--- xserver-xorg-input-synaptics.orig/src/eventcomm.c	2011-07-05 11:55:08.771347676 +0300
+++ xserver-xorg-input-synaptics/src/eventcomm.c	2011-07-05 11:55:10.801347676 +0300
@@ -468,6 +468,7 @@
                            XI_TouchEnd, 0, ecpriv->touch_mask);
         ecpriv->mt_slot_map[ecpriv->cur_slot] = -1;
         ecpriv->close_slot = FALSE;
+        ecpriv->active_touches--;
     }
     else
     {
@@ -479,9 +480,12 @@
             int y = valuator_mask_get(ecpriv->touch_mask, y_axis);
 
             if (is_inside_active_area(priv, x, y))
+            {
                 xf86PostTouchEvent(pInfo->dev,
                                    ecpriv->mt_slot_map[ecpriv->cur_slot],
                                    XI_TouchBegin, 0, ecpriv->touch_mask);
+                ecpriv->active_touches++;
+            }
             else
                 ecpriv->mt_slot_map[ecpriv->cur_slot] = -1;
         }
@@ -540,7 +544,9 @@
     case EV_SYN:
         switch (ev->code) {
         case SYN_REPORT:
-            if (comm->oneFinger)
+            if (priv->has_touch)
+                hw->numFingers = ecpriv->active_touches;
+            else if (comm->oneFinger)
                 hw->numFingers = 1;
             else if (comm->twoFingers)
                 hw->numFingers = 2;
Index: xserver-xorg-input-synaptics/src/eventcomm.h
===================================================================
--- xserver-xorg-input-synaptics.orig/src/eventcomm.h	2011-07-05 11:55:02.771347676 +0300
+++ xserver-xorg-input-synaptics/src/eventcomm.h	2011-07-05 11:55:10.801347676 +0300
@@ -58,6 +58,7 @@
     int num_touches;
     struct mtdev *mtdev;
     struct grail *grail;
+    int active_touches;
 } EventcommPrivate;
 
 extern Bool EventProcessEvent(InputInfoPtr pInfo, struct CommData *comm,
